Dockerfile

```
@@ -28,45 +21,71 @@ ENV PYTHON_PIPENV_VERSION=2024.1.0
-# Python virtual environment used in the final image.
-RUN python3 -m pip install --break-system-packages --no-cache-dir --upgrade pipenv==${PYTHON_PIPENV_VERSION} \
-  # Manually create Python virtual environment for the final image
-  && python3 -m venv ${VIRTUAL_ENV} \
-  # Ensure the core Python packages are installed in the virtual environment
-  && ${VIRTUAL_ENV}/bin/python3 -m pip install --no-cache-dir --upgrade \
-    pip==${PYTHON_PIP_VERSION} \
-    setuptools==${PYTHON_SETUPTOOLS_VERSION} \
-    wheel==${PYTHON_WHEEL_VERSION}
+###
+# Create a Python virtual environment (venv) for setup (due to PEP 668); install the
+# specified versions of pip, setuptools, and wheel into the setup venv; install the
+# specified version of pipenv into the setup venv; create the image dependency venv;
+# and install the specified versions of pip, setuptools, and wheel into the dependency
+# venv.
+#
+# Note that we use the --no-cache-dir flag to avoid writing to a local
+# cache.  This results in a smaller final image, at the cost of
+# slightly longer install times.
+###
+RUN python3 -m venv --system-site-packages /usr/local \
+    # Ensure the core Python packages are installed in the virtual environment
+    && /usr/local/bin/python3 -m pip install --no-cache-dir --upgrade \
+        pip==${PYTHON_PIP_VERSION} \
+        setuptools==${PYTHON_SETUPTOOLS_VERSION} \
+        wheel==${PYTHON_WHEEL_VERSION} \
+    && /usr/local/bin/python3 -m pip install --no-cache-dir --upgrade \
+        pipenv==${PYTHON_PIPENV_VERSION} \
+    # Manually create the virtual environment
+    && python3 -m venv --system-site-packages ${VIRTUAL_ENV} \
+    # Ensure the core Python packages are installed in the virtual environment
+    && ${VIRTUAL_ENV}/bin/python3 -m pip install --no-cache-dir --upgrade \
+        pip==${PYTHON_PIP_VERSION} \
+        setuptools==${PYTHON_SETUPTOOLS_VERSION} \
+        wheel==${PYTHON_WHEEL_VERSION}
 
-# Install code-gov-update Python requirements
+###
+# Check the Pipfile configuration and then install the Python dependencies into
+# the virtual environment.
+#
+# Note that pipenv will install into a virtual environment if the VIRTUAL_ENV
+# environment variable is set.
+###
 WORKDIR /tmp
 COPY src/Pipfile src/Pipfile.lock ./
-RUN pipenv sync --clear --verbose
+RUN pipenv check --verbose \
+    && pipenv install --clear --deploy --extra-pip-args "--no-cache-dir" --verbose
 
 # The version of Python used here should match the version of the Alpine
 # python3 package installed in the compile-stage.
-FROM python:3.12.7-alpine3.20 as build-stage
+FROM docker.io/library/python:3.12.8-alpine3.21 AS build-stage
+
+###
+# For a list of pre-defined annotation keys and value types see:
+# https://github.com/opencontainers/image-spec/blob/master/annotations.md
+#
+# Note: Additional labels are added by the build workflow.
+###
+LABEL org.opencontainers.image.authors="vm-fusion-dev-group@trio.dhs.gov"
```

We should update this email address while we are here.
```suggestion
LABEL org.opencontainers.image.authors="vm-dev@gwe.cisa.dhs.gov"
```

---

Do you mind if I slap this into another PR? I have at least one (maybe two) followup PRs already planned and I've already been forced to overload the changes in this PR to get tests to pass.

---

Fine by me, I just figured it was easy to do in here and didn't want it to slip thru the cracks.

---

Please see https://github.com/cisagov/code-gov-update/pull/179.

---

Thanks for taking care of that!